src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ProducerConsumer.cs

```diff
@@ -353,29 +369,52 @@ async Task<TResult> ReadFromChannelAndConsumeItemsAsync()
             return await consumeItems(channel.Reader, args, cancellationToken).ConfigureAwait(false);
         }
 
-        async Task ProduceItemsAndWriteToChannelAsync()
```

Comment:
the original code in here moved to PerformActionAndCloseWriterAsync so that it could be used by the new helper i added recently.